### DISCUSSION TITLE
Stop Program and Image Display Queries

### DISCUSSION THREAD

**Discussion by MarieugiMR** [bonsai-rx/bonsai#1375] [Q&A]
Hi, I recently started using the Bonsai system, and I'm extremely excited about it. I'm about to begin an fMRI experiment where I've created a code that changes the stimulus (in this case, images) every time a specific key is pressed. The code works perfectly for the requirements of this experiment. However, there are two things I'm unsure about:

a. How can I instruct the program to stop when there is no key press for 4 seconds, essentially at the end of the presentation? I've tried various approaches, such as incorporating "Flip" or experimenting with "Threshold" and "Timer," but none of them have yielded the desired results.

b. I would like to include black or grey images at the beginning for 4 seconds. Following that, the subsequent images should be displayed, with every second image being followed by another grey or black image. Finally, at the end of the experiment, which lasts for each scan's duration of 2 minutes and 45 seconds, I want to display a grey or black image once again.

I hope this explanation clarifies my requirements. Thank you for your assistance.

**Comment by bruno-f-cruz** [comment]
Can you include your current workflow file? thanks

**Comment by MarieugiMR** [comment]
sure1 Thank you. Please find the workflow file attached. Also, I want to
add a black dot or a small cross in the middle of each photo, serving as a
fixation spot. How do I add this? As an overlapped image?



Thank you in advance.

**Reply by bruno-f-cruz** [reply]
There is no attached file.

**Comment by MarieugiMR** [comment]
I can see the attachment from my end, but I am re-attaching it here as
well. Hope this time it will work.

**Reply by bruno-f-cruz** [reply]
Heya, 
can you attach the file as suggested in #864 ? Also, I edited your post to prevent the webpage from loading all the text. Thanks!

**Comment by MarieugiMR** [comment]
Thank you very much. I just uploaded the zip file over there.

**Reply by bruno-f-cruz** [reply]
Sorry for the confusion, but what I meant was that you should attach the file as written in #864,  but in this post :). You should definitely NOT use the #864 discussion to ask your question.
Do you mind copy-pasting it to this discussion? (I can delete the other one once you copy it here no worries).

**Comment by MarieugiMR** [comment]
it doesn allow me to send the files. So I zipped only 10 out of the 100
images and I added the workflow too.

**Reply by bruno-f-cruz** [reply]
There is nothing attached to your post. Am I missing something?

**Comment by MarieugiMR** [comment]
That's interesting. The attachment is there. I can see that. I will try to
reattach it here. Let me know if that worked. If not, is there any other
way for me to send it to you?

**Reply by bruno-f-cruz** [reply]
Looking at your messages, it seems you are trying to answer this thread via email. Can you try logging in on GitHub directly, and simply drag and drop the .zip file to the message you are writing?
If you have trouble, please take a look at [GitHub docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files)

**Reply by MarieugiMR** [reply]
[TEST_BONSAI.zip](https://github.com/bonsai-rx/bonsai/files/12050186/TEST_BONSAI.zip)

Oh!I see! I hope that now it worked.

**Reply by bruno-f-cruz** [reply]
`a. How can I instruct the program to stop when there is no key press for 4 seconds, essentially at the end of the presentation? I've tried various approaches, such as incorporating "Flip" or experimenting with "Threshold" and "Timer," but none of them have yielded the desired results.`

You can use `Throttle` to achieve this logic. Then all you need to do is to close the sequence with the first value.
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/038bf0b9-eee6-4272-bf35-75f6ecc743f0)

`b. I would like to include black or grey images at the beginning for 4 seconds. Following that, the subsequent images should be displayed, with every second image being followed by another grey or black image. Finally, at the end of the experiment, which lasts for each scan's duration of 2 minutes and 45 seconds, I want to display a grey or black image once again.`

I would start with the following tutorial https://bonsai-rx.org/docs/tutorials/state-machines.html
Then your trial structure could be implemented inside each `SelectMany` "state". Give it a try and let me know if you need more help!

Best,
Bruno

**Reply by MarieugiMR** [reply]
Thanks Bruno. Also, I want to
add a black dot or a small cross in the middle of each photo, serving as a
fixation spot. How do I add this? As an overlapped image?

**Comment by MarieugiMR** [comment]
Hi Bruno, I don’t believe i’ve made progress, but I wanted to reach out
before considering giving up. I still can’t accomplish waht you suggested,
and i also have another question. In fMRI experiments with MATLAB,I
typically receive matfiles that provide experiment parameters. How does
this work with bonsai?

**Reply by bruno-f-cruz** [reply]
I will need a bit more information to be able to help you. With what question do you need help with exactly? What kind of settings do you want to change and how?
